### DISCUSSION TITLE
How to obtain character range of null value in map

### DISCUSSION THREAD

**Discussion by easbar** [eemeli/yaml#231] [Q&A]
```javascript
YAML.parseDocument(`name:      \nage: 26`).contents.items
```
is a list of objects of type 'PAIR' that have a key and value property.
Each key and value provide a `range` field that is very useful to figure out the exact location in the original text. Unfortunately, for `name` in this example the values is simply `null` so it is not possible to obtain the range. How can we get this range?

I tried 

```javascript
YAML.parseDocument(`name:      \nage: 26`, {keepCstNodes: true}).contents.cstNode.items
```
 which seems to be an alternating list of objects of type `PLAIN` and `MAP_VALUE` which provide a `range` field even for the 'null' value of `name`. The range we get for the `MAP_VALUE` corresponding to `name` is exactly what I am looking for. But unfortunately if I write the same YAML as 

```javascript
const doc = YAML.parseDocument(`{name:     , age: 26}`, {keepCstNodes: true});
```
`doc.contents.cstNode.items` will be an array of some objects with type `PLAIN` (for `name`, `age` and `26`) and some nodes that simply provide a `character` and `offset` (for `{`, `:`, `,` and `}`). So this does not seem to be a reliable way to obtain the character range of the `name` value (in case it is null).

My use-case is that I want to provide auto-completion for the `name` field and for this I need to detect the position of the name value in the YAML document. My current workaround is basing the value range on the range of the key, but that is not ideal either.

**Comment by eemeli** [comment]
You've managed to hit one of the issues (the handling of empty nodes) that lead me to rewrite the parser for `yaml@2`. Admittedly this particular case could (and probably should) be handled by the v1 AST composer as well, but, well, it clearly doesn't.

The new parser has been merged into the `master` branch, but I haven't yet cut a beta release from it due to a few missing features & incomplete documentation. It does report a range for your input, albeit placing it just after all the whitespace, including the newline:

```js
const doc = YAML.parseDocument(`name:      \nage: 26`)
doc.contents.items[0].value
// Scalar { value: null, range: [ 12, 12 ], source: '', type: 'PLAIN' }
```

That range ought to be `[ 11, 11 ]`.

**Reply by easbar** [reply]
Thanks a lot for your quick reply (and this nice parser of course) :) So this means the new parser changes `doc.contents.items[0].value` from `null` to the `Scalar` object you wrote as a comment? That's great to hear and certainly seems more consistent. But why is the character range `[11, 11]` and does not start right after `name:`? And if not would it be reliable to assume that everything between the `name` key` and this range was whitespace?

Btw I ran into another, similar (but different) case (using the version `2.0.0-3`):

```javascript
const doc = YAML.parseDocument(`people: [{age: 16, name:`)
console.log(doc.contents.items[0].value.items[0].items[1].value);
// Scalar {value: null, sourceStr: "", range: Array(2), cstNode: PlainValue, type: "PLAIN"}
```

Here the value is not null. I guess, because the 'empty' value of name is at the very end of the input string.

**Reply by easbar** [reply]
I see you adjusted the range in ed3bce795a073402a000bfc65e14d944c440bae6, thanks a lot for this. So for my initial example we now get `Scalar { value: null, range: [ 11, 11 ], source: '', type: 'PLAIN' }`. But can you explain why the range is 'empty'? Should it not rather be [11, 12]? In this case it would be the range of the new line character. But why does it not just span all the whitespace between the colon after `name` and the next non-empty node (`age`)?

**Reply by eemeli** [reply]
That would be because a string consisting of one or more spaces is not a valid representation of an empty node. Technically, the YAML spec-mandated position for the empty node would be immediately after the `:` character, but I figured that it'd be more useful to place it at the end of the whitespace, as there it's possible to actually enter a non-empty value.

It is possible for there to be a tag or an alias between the `:` and the position of the empty string. And possibly comments as well, if they're followed by a tag or alias.
